File: [2023.09.25 - 2023.10.01] - 4주차/김유진/섬 연결하기.java
import java.util.*;

class Cost implements Comparable<Cost>{
    int i1; 
    int i2;
    int cost;
    
    public Cost(int i1, int i2, int cost){
        this.i1=i1; 
        this.i2=i2;
        this.cost=cost;
    }
    
    public int compareTo(Cost c){
        return this.cost-c.cost;
    }
}

class Solution {
    public int solution(int n, int[][] costs) {
        
        PriorityQueue<Cost> queue=new PriorityQueue();
        for(int[] cost: costs){
            queue.add(new Cost(cost[0], cost[1], cost[2]));
        }
        
        //Parent Array
        int[] parents=new int[n];
        for(int i=0;i<n;i++){
            parents[i]=i;
        }
        
        int sum=0;
        while(!queue.isEmpty()){
            Cost nowCost=queue.poll();
            
            if(isPossible(nowCost, parents)){
                combine(nowCost, parents);
                sum+=nowCost.cost;
            }
        }
        
        return sum;
    }
    
    private void combine(Cost cost, int[] parents){
        int i1=cost.i1; int i2=cost.i2;
        int p1=getParent(i1, parents); int p2=getParent(i2, parents);
        
        parents[p2]=p1;
    }
    
    private boolean isPossible(Cost cost, int[] parents){
        int i1=cost.i1; int i2=cost.i2;
        int p1=getParent(i1, parents); int p2=getParent(i2, parents);
        
        if(p1==p2)
            return false;
        
        return true;
    }
    
    private int getParent(int island, int[] parents){
        
        if(island==parents[island])
            return island;
        
        return getParent(parents[island], parents);
    }
}

File: [2023.09.18 - 2023.09.24] - 3주차/박민형/섬 연결하기.py
# 1. 크루스칼 알고리즘 사용(최소 비용으로 그래프를 연결할 때 사용)

def solution(n, costs):
    answer = 0
    costs.sort(key=lambda x: x[2])
    link = set([costs[0][0]])

    while len(link) < n:
        for cost in costs:
            # 이미 최소비용으로 두 섬간 연결이 되어있는 경우
            if cost[0] in link and cost[1] in link:
                continue
            # 한 섬만 연결되어 있는 경우
            if cost[0] in link or cost[1] in link:
                link.update([cost[0], cost[1]])
                answer += cost[2]
                break

    return answer